python/lib/communication/dmod/communication/metadata_message.py

```diff
@@ -0,0 +1,177 @@
+from .message import AbstractInitRequest, MessageEventType, Response
+from enum import Enum
+from numbers import Number
+from typing import Dict, Optional, Union
+
+
+class MetadataPurpose(Enum):
+    CONNECT = 1,
+    """ The metadata relates to the opening of a connection. """
+    DISCONNECT = 2,
+    """ The metadata relates to the closing of a connection. """
+    PROMPT = 3,
+    """ The metadata it a prompt to the other side that it should send any needed metadata messages. """
+    UNCHANGED = 4,
+    """ The metadata indicate nothing needs to be communicated in response to a ``PROMPT``. """
+    CHANGE_CONFIG = 5
+    """ The metadata indicate something regarding the config of how the other side behaves needs changing. """
+
+    @classmethod
+    def get_value_for_name(cls, name_str: str) -> Optional['MetadataPurpose']:
+        cleaned_up_str = name_str.strip().lower()
+        for value in cls:
+            if value.name.lower() == cleaned_up_str:
+                return value
+        return None
+
+
+class MetadataMessage(AbstractInitRequest):
+
+    event_type: MessageEventType = MessageEventType.METADATA
+
+    _purpose_serial_key = 'purpose'
+    _description_serial_key = 'description'
+    _metadata_follows_serial_key = 'additional_metadata'
+    _config_changes_serial_key = 'config_changes'
+    _config_change_dict_type_key = 'config_value_dict_type'
+
+    @classmethod
+    def factory_init_from_deserialized_json(cls, json_obj: dict) -> Optional['MetadataMessage']:
+        if cls._purpose_serial_key not in json_obj:
+            return None
+        purpose = MetadataPurpose.get_value_for_name(json_obj[cls._purpose_serial_key])
+        if purpose is None:
+            return None
+        if cls._metadata_follows_serial_key in json_obj:
+            metadata_follows = json_obj[cls._metadata_follows_serial_key]
+        else:
+            # default to False for this, as this is pretty safe assumption if we don't see it explicit
+            metadata_follows = False
+        description = json_obj[cls._description_serial_key] if cls._description_serial_key in json_obj else None
+        cfg_changes = json_obj[cls._config_changes_serial_key] if cls._config_changes_serial_key in json_obj else None
+        return cls(purpose=purpose, description=description, metadata_follows=metadata_follows,
+                   config_changes=cfg_changes)
+
+    @classmethod
```

Comment:
The intent for this is to be able to grant read-only access to a standard key for use during serialization/deserialization in some circumstances.  It'll get used within an instance-specific dictionary object,  but in times and places it is needed, it should always be the same.  So, I think a static class attribute makes the most sense.

Reply:
Well `@property` is "read-only" unless you provide an explicit `@config-change_dict_type_key.setter` function.  Either way works for me.  We just want to keep it consistent.

Reply:
At least on the surface, having an instance property _seems_ like it could make it easier for the expected/utilized key in various places to not be globally in sync.  It might be that this isn't really the case, especially in Python (i.e., protections might actually just be assumed _conventions_).  Even then, a class attribute has a slightly lower memory footprint, which is a tangible albeit very small advantage.

I don't immediately see advantages to an instance property, though I'm not opposed to changing things that way if there are some (even if it is something non-functional, like being generally more intuitive to follow).

Reply:
Not enough to worry about for the moment.